harness/determined/keras/_tf_keras_trial.py

```diff
@@ -26,49 +29,52 @@
-                        "Error when creating the weights of optimizer, making it "
-                        "impossible to restore the saved optimizer state. As a result, "
-                        "your model is starting with a freshly initialized optimizer."
-                    )
-            else:
-                # Build train function (to get weight updates).  Models that aren't
-                # graph networks must wait until they are called with data to
-                # _make_train_function() and so can't load optimizer weights.
-                model._make_train_function()
+    tf2_2_or_newer = version.parse(tf.__version__) >= version.parse("2.2.0")
+    if model._is_graph_network or tf2_2_or_newer:  # pylint: disable=protected-access
+        if not optimizer:
+            optimizer = model.optimizer
 
-            optimizer_weight_values = load_optimizer_weights_from_hdf5_group(f)
+        if tf2_2_or_newer:
             try:
-                model.optimizer.set_weights(optimizer_weight_values)
-            except ValueError:
+                optimizer._create_all_weights(model.trainable_variables)
+            except (NotImplementedError, AttributeError):
                 logging.warning(
-                    "Error in loading the saved optimizer "
-                    "state. As a result, your model is "
-                    "starting with a freshly initialized "
-                    "optimizer."
+                    "Error when creating the weights of optimizer, making it "
+                    "impossible to restore the saved optimizer state. As a result, "
+                    "your model is starting with a freshly initialized optimizer."
                 )
         else:
+            # Build train function (to get weight updates).  Models that aren't
+            # graph networks must wait until they are called with data to
+            # _make_train_function() and so can't load optimizer weights.
+            model._make_train_function()
+
+        optimizer_weight_values = load_optimizer_weights_from_hdf5_group(h5group)
+        try:
+            optimizer.set_weights(optimizer_weight_values)
+        except ValueError:
             logging.warning(
-                "Sequential models without an `input_shape` "
-                "passed to the first layer cannot reload their "
-                "optimizer state. As a result, your model is "
-                "starting with a freshly initialized optimizer."
+                "Error in loading the saved optimizer "
+                "state. As a result, your model is "
+                "starting with a freshly initialized "
+                "optimizer."
             )
+    else:
+        logging.warning(
+            "Sequential models without an `input_shape` "
+            "passed to the first layer cannot reload their "
+            "optimizer state. As a result, your model is "
+            "starting with a freshly initialized optimizer."
+        )
```

Comment:
Seems like we should detect this case at the beginning of training, rather than at the beginning of continuing training.

Reply:
separately, is this changing the scope of types of models that will have issues here?  Did sequential models in tf1.15 really not work here?

Reply:
Ok, I still think this code is super confusing, but I do see that it hasn't changed.  I still don't see how:

```python
if model._is_graph_network or tf2_2_or_newer:
    ...
else:
    # we have a sequential model without 'input_shape' as the first layer
```

Is that what `model._is_graph_network` means?

Reply:
Based on [this](https://github.com/tensorflow/tensorflow/blob/590d6eef7e91a6a7392c8ffffb7b58f2e0c8bc6b/tensorflow/python/keras/saving/hdf5_format.py#L196) I would have to say yes.